src/cosl/rules.py

```diff
@@ -262,82 +261,98 @@ def _from_file(  # noqa: C901
-                            query_type=self.query_type,
-                        )
-
             return groups
 
-    def _group_name(self, root_path: str, file_path: str, group_name: str) -> str:
-        """Generate group name from path and topology.
-
-        The group name is made up of the relative path between the root dir_path, the file path,
-        and topology identifier.
+    def _from_str(
+        self,
+        yaml_str: Dict[str, Any],
+        *,
+        group_name: Optional[str] = None,
+        group_name_prefix: Optional[str] = None,
+    ) -> List[OfficialRuleFileItem]:
+        """Process rules from string, injecting juju topology. If a single-rule format is provided, a hash of the yaml file is injected into the group name to ensure uniqueness.
 
         Args:
-            root_path: path to the root rules dir.
-            file_path: path to rule file.
-            group_name: original group name to keep as part of the new augmented group name
+            yaml_str: rules content in single-rule or official-rule format as a YAML dict
+            group_name: a custom identifier for the rule name to include in the group name
+            group_name_prefix: a custom group identifier to prefix the resulting group name, likely Juju topology and relative path context
 
-        Returns:
-            New group name, augmented by juju topology and relative path.
+        Raises:
+            ValueError, when invalid rule format given.
         """
-        rel_path = os.path.relpath(os.path.dirname(file_path), root_path)
-        rel_path = "" if rel_path == "." else rel_path.replace(os.path.sep, "_")
+        if not yaml_str:
+            raise ValueError("Empty")
+
+        if self._is_official_rule_format(yaml_str):
+            yaml_str = yaml_str
+            groups = yaml_str["groups"]
+        elif self._is_single_rule_format(yaml_str, self.rule_type):
+            yaml_str = yaml_str
+            if not group_name:
+                # Note: the caller of this function should ensure this never happens:
+                # Either we use the standard format, or we'd pass a group_name.
+                # If/when we drop support for the single-rule-per-file format, this won't
+                # be needed anymore.
+                group_name = hashlib.shake_256(str(yaml_str).encode("utf-8")).hexdigest(10)
+            else:
+                group_name = self._sanitize_metric_name(group_name)
 
-        # Generate group name:
-        #  - name, from juju topology
-        #  - suffix, from the relative path of the rule file;
-        group_name_parts = [self.topology.identifier] if self.topology else []
-        group_name_parts.extend([rel_path, group_name, f"{self.rule_type}s"])
-        # filter to remove empty strings
-        return "_".join(filter(None, group_name_parts))
+            # convert to list of groups to match official rule format
+            groups = [{"name": group_name, "rules": [yaml_str]}]
```

Comment:
I assume you mean prefix not suffix and I dont think this is a good suggestion because later on in the code we:

``` python
if not self._is_already_modified(group["name"]):
    # update group name with topology and sub-path
    group["name"] = "_".join(filter(None, [group_name_prefix, group["name"], f"{self.rule_type}s"]))
```

Which prefixes the group_name_prefix otherwise we would have `prefix_prefix_name_alerts`